File: python/segment.py
#!/usr/bin/env python3
"""AI segmentation sidecar — writes an 8-bit grayscale mask PNG (white = selected).

Companion to the Rust bridge (src/segment.rs), following the same sidecar
pattern as denoise.py: the Rust side shells out, this script does one job and
exits non-zero with a human-readable reason on stderr when it can't.

Usage:
  python segment.py --input photo.png --output mask.png --target subject|sky

Backends (weights auto-download to the USER's home cache on first run — nothing
is stored in the repo, consistent with .gitignore'ing python/weights):
  subject -> rembg, U^2-Net salient-object alpha  (pip install rembg; ~/.u2net)
  sky     -> SegFormer-B0 fine-tuned on ADE20K via transformers
             (pip install transformers torch; ~/.cache/huggingface)

The output mask is soft (the model's own alpha / class probability), which the
render engine samples bilinearly — so edges come pre-feathered.
"""

import argparse
import sys


def die(msg: str) -> None:
    print(f"segment.py: {msg}", file=sys.stderr)
    sys.exit(2)


def subject_mask(img_path: str):
    """Salient-subject alpha via rembg's default U^2-Net session."""
    try:
        from rembg import remove
    except ImportError:
        # ASCII-only: Windows consoles in legacy codepages mangle wide dashes.
        die(
            "subject segmentation needs rembg -> pip install rembg "
            "(the U^2-Net model auto-downloads to ~/.u2net on first run)"
        )
    from PIL import Image

    img = Image.open(img_path).convert("RGB")
    # only_mask=True returns the soft alpha as a single-channel PIL image.
    return remove(img, only_mask=True)


def sky_mask(img_path: str):
    """ADE20K semantic segmentation, sky-class probability as the mask."""
    try:
        import torch
        from transformers import (
            SegformerForSemanticSegmentation,
            SegformerImageProcessor,
        )
    except ImportError:
        # ASCII-only: Windows consoles in legacy codepages mangle wide dashes.
        die(
            "sky segmentation needs transformers + torch -> pip install transformers "
            "(SegFormer-B0 ADE20K, ~14 MB, auto-downloads to ~/.cache/huggingface)"
        )
    import numpy as np
    from PIL import Image

    name = "nvidia/segformer-b0-finetuned-ade-512-512"
    processor = SegformerImageProcessor.from_pretrained(name)
    model = SegformerForSemanticSegmentation.from_pretrained(name)
    model.eval()

    # Resolve the sky class from the model's own label table instead of
    # hard-coding an index — survives label-map revisions.
    sky_ids = [i for i, l in model.config.id2label.items() if "sky" in l.lower()]
    if not sky_ids:
        die(f"model {name} has no 'sky' class in id2label — cannot build a sky mask")

    img = Image.open(img_path).convert("RGB")
    with torch.no_grad():
        inputs = processor(images=img, return_tensors="pt")
        logits = model(**inputs).logits  # (1, n_classes, h/4, w/4)
        up = torch.nn.functional.interpolate(
            logits, size=(img.height, img.width), mode="bilinear", align_corners=False
        )
        probs = up.softmax(dim=1)[0]

    m = probs[int(sky_ids[0])].numpy()
    m = (m * 255.0).clip(0, 255).astype(np.uint8)
    return Image.fromarray(m, mode="L")


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--input", required=True, help="source image (any PIL-readable format)")
    ap.add_argument("--output", required=True, help="mask PNG to write (8-bit grayscale)")
    ap.add_argument("--target", required=True, choices=["subject", "sky"])
    a = ap.parse_args()

    mask = subject_mask(a.input) if a.target == "subject" else sky_mask(a.input)
    mask = mask.convert("L")
    mask.save(a.output)
    print(f"segment.py: {a.target} mask -> {a.output}")


if __name__ == "__main__":
    main()

File: src/segment.rs
//! AI segmentation bridge — Rust side of the sidecar (`python/segment.py`).
//!
//! Same shell-out pattern as [`crate::denoise`] (SCUNet): a local Python
//! process does the model inference and writes an 8-bit grayscale mask PNG
//! (white = selected, soft edges = the model's own alpha), which the app then
//! attaches to the recipe as a [`crate::recipe::MaskGeometry::Bitmap`] local
//! adjustment. The AI picks *where*; every actual edit stays a deterministic
//! recipe slider. Models auto-download to the user's home caches on first run
//! (`~/.u2net`, `~/.cache/huggingface`) — no weights in the repo.

use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use anyhow::{bail, Context, Result};

use crate::config::Config;

/// Everything one segmentation run needs; built from [`Config`] like
/// [`crate::denoise::DenoiseOpts`].
pub struct SegmentOpts {
    pub python_bin: String,
    pub script: PathBuf,
    /// `"subject"` (U²-Net salient object) or `"sky"` (SegFormer ADE20K).
    pub target: String,
}

impl SegmentOpts {
    pub fn from_config(cfg: &Config, target: &str) -> Self {
        SegmentOpts {
            python_bin: cfg.python_bin.clone(),
            script: PathBuf::from(&cfg.segment_script),
            target: target.to_string(),
        }
    }
}

/// Run the sidecar: `input` (any image file) → `output` (8-bit grayscale PNG).
/// The mask is in the INPUT's frame — feed it the original-frame preview so it
/// lands in the same space recipe masks live in.
pub fn segment_file(opts: &SegmentOpts, input: &Path, output: &Path) -> Result<()> {
    if !opts.script.exists() {
        bail!(
            "segmentation sidecar not found at {} — run from the project dir or set \
             AUTOSHOP_SEGMENT_SCRIPT.",
            opts.script.display()
        );
    }
    crate::pipeline::ensure_parent(output)?;
    let mut cmd = Command::new(&opts.python_bin);
    cmd.arg(&opts.script)
        .arg("--input")
        .arg(input)
        .arg("--output")
        .arg(output)
        .arg("--target")
        .arg(&opts.target)
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit());
    // Don't flash a console window when the windowed GUI spawns the sidecar.
    crate::hide_child_console(&mut cmd);
    let status = cmd.status().with_context(|| {
        format!(
            "launch segmentation sidecar ({} {}) — is Python on PATH / AUTOSHOP_PYTHON set?",
            opts.python_bin,
            opts.script.display()
        )
    })?;
    if !status.success() {
        bail!(
            "segmentation sidecar exited with {} (see its log above — a missing \
             dependency prints the exact pip install line)",
            status.code().map(|c| c.to_string()).unwrap_or_else(|| "signal".into())
        );
    }
    if !output.exists() {
        bail!("sidecar reported success but wrote no mask at {}", output.display());
    }
    Ok(())
}
